ccm_web/server/src/api/api.enroll.section.users.handler.ts

```diff
@@ -0,0 +1,99 @@
+import CanvasRequestor from '@kth/canvas-api'
+
+import { APIErrorData, isAPIErrorData } from './api.interfaces'
+import { handleAPIError, HttpMethod } from './api.utils'
+
+import baseLogger from '../logger'
+import { CanvasEnrollment } from '../canvas/canvas.interfaces'
+import { SectionUserDto } from './dtos/api.section.users.dto'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+export class EnrollSectionUsersApiHandler {
+  requestor: CanvasRequestor
+  users: SectionUserDto[]
+  sectionId: number
+
+  constructor (requestor: CanvasRequestor, users: SectionUserDto[], sectionId: number) {
+    this.requestor = requestor
+    this.users = users
+    this.sectionId = sectionId
+  }
+
+  makeResponse (enrollmentResult: Array<APIErrorData | CanvasEnrollment>): CanvasEnrollment[] | APIErrorData {
+    const failures = []
+    const statusCodes: Set<number> = new Set()
+    const successes = []
+
+    for (const enrollment of enrollmentResult) {
+      if (isAPIErrorData(enrollment)) {
+        const {
+          statusCode,
+          errors
+        } = enrollment
+        failures.push(...errors)
+        statusCodes.add(statusCode)
+      } else {
+        successes.push(enrollment)
+      }
+    }
+
+    if (successes.length === this.users.length) {
+      return successes
+    } else {
+      return {
+        statusCode: statusCodes.size > 1 ? 502 : [...statusCodes][0],
+        errors: failures
+      }
+    }
+  }
+
+  async enrollUser (user: SectionUserDto): Promise<CanvasEnrollment | APIErrorData> {
+    try {
+      const endpoint = `sections/${this.sectionId}/enrollments`
+      const method = HttpMethod.Post
+      const body = {
+        enrollment: {
+          // 'sis_login_id:' prefix per...
+          // https://canvas.instructure.com/doc/api/file.object_ids.html
+          user_id: `sis_login_id:${user.loginId}`,
```

Comment:
Is this where we are doing something assuming the client sent uniqnames?  It's now unclear to me what the expectation is in that regard.  There is talk of the user including non-umich emails in the csv and *somewhere* there is expected to be an automatic translation to append +gmail.com.

Reply:
I think the automatic translation was `@` to `+`, maybe that's what you meant. Looking at the Slack discussion, I think maybe there was confusion on this point. I think Canvas just expects the `@` sign to be a `+`, for whatever reason. You're right in that we can't know what's after the `@`.

Reply:
so the user could put in a user id of 'rowc' (@umich.edu) or 'rowc@gmail.com' and the client should send 'rowc' and 'rowc+gmail.com' respectively, and either way the above code is correct?

Reply:
I think so. A note (which you may know), I don't think we would ever want to send `@umich.edu`. I think either the loginId would be a simple string OR a non-`umich` email address. I think what's important is that it matches what is in Canvas, and UM users wouldn't have the email ending.

Reply:
I'm using the `sis_login_id` prefix here, as directed by @pushyamig.  It enables references to the user by their Canvas "login ID".  I don't know about all users, but in my case, my login ID is my uniqname, "lsloan", and nothing else, not my email address.  Likewise, the test users I've worked with have login IDs "canvasa", "canvasb", etc., like uniqnames, not email addresses.  When I do the test of 400 users, I use a file of STATS 250 student uniqnames, not email addresses.

I saw the conversation about email addresses ("xxx@umich.edu") and some other kind of ID ("xxx+umich.edu"), but that didn't seem to apply to this situation, so I don't know the details about those.

This code doesn't do any validation of the names used because there had been discussion to let Canvas return errors about missing/improper/inaccessible users and this API would return those to the client.  That also solves the validation problem with our organization using names of 3–8 lower-case alphabetic characters while other organizations may use different criteria.

Reply:
I've added code to handle email addresses as SIS login ID values.